subprojects/logging/src/main/java/org/gradle/internal/logging/console/AbstractUserInputRenderer.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.logging.console;
+
+import org.gradle.internal.logging.events.OutputEvent;
+import org.gradle.internal.logging.events.OutputEventListener;
+import org.gradle.internal.logging.events.UserInputRequestEvent;
+import org.gradle.internal.logging.events.UserInputResumeEvent;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ListIterator;
+
+public abstract class AbstractUserInputRenderer implements OutputEventListener {
+
+    final OutputEventListener delegate;
+    private final List<OutputEvent> eventQueue = new ArrayList<OutputEvent>();
+    private boolean paused;
+
+    public AbstractUserInputRenderer(OutputEventListener delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public void onOutput(OutputEvent event) {
+        if (event instanceof UserInputRequestEvent) {
+            String prompt = ((UserInputRequestEvent) event).getPrompt();
+            handleUserInputRequestEvent(prompt);
+            return;
+        } else if (event instanceof UserInputResumeEvent) {
+            handleUserInputResumeEvent();
+            return;
+        }
+
+        if (paused) {
+            eventQueue.add(event);
+            return;
+        }
+
+        delegate.onOutput(event);
+    }
+
+    private void handleUserInputRequestEvent(String prompt) {
+        startInput(prompt);
+        paused = true;
+    }
+
+    private void handleUserInputResumeEvent() {
+        if (!paused) {
+            throw new IllegalStateException("Cannot resume user input if not paused yet");
+        }
+
+        paused = false;
+        finishInput();
+        replayEvents();
+    }
+
+    private void replayEvents() {
+        ListIterator<OutputEvent> iterator = eventQueue.listIterator();
+
+        while (iterator.hasNext()) {
```

Comment:
⭕️ You could use the `Queue` interface for this.

Reply:
I believe for that to work, `OutputEvent` would have to implement `java.lang.Comparable`. Not going to do that for now.

Reply:
I'm not sure what you mean. You can use an ArrayDeque and use take() instead of get() and remove().

Reply:
Fair enough, using `ArrayDeque`.

Reply:
Had to revert that as some integration tests failed due to Java 5 compatibility.